run_tests.py.in

```diff
@@ -0,0 +1,665 @@
+            '--track-origins=yes',
+            '--num-callers=50',
+            '--leak-check=full',
+            '--show-reachable=yes',
+            '--malloc-fill=0xa1',
+            '--free-fill=0xa1',
+        ]
+    else:
+        _ctx.valgrind_opts = []
+
+    if wine:
+        _ctx.wine = shutil.which('wine')
+        _ctx.re2c += '.exe'
+        # Ignore whitespace at the end of line
+        _ctx.diff += ' -b'
+    elif os.name == 'nt':
+        # This should find 're2c.exe' in os.getcwd()
+        _ctx.re2c += shutil.which('re2c', path=os.getcwd())
+        # Ignore whitespace at the end of line
+        _ctx.diff += ' -b'
+
+
+def environment_check(valgrind=False, wine=False):
+    """Check environment and requirements."""
+    if not shutil.which('diff'):
+        raise RuntimeError("The 'diff' utility is missing from PATH")
+
+    if valgrind and not shutil.which('valgrind'):
+        raise RuntimeError("The 'valgrind' tool is missing from the PATH")
+
+    if wine and not shutil.which('wine'):
+        raise RuntimeError("The 'wine' layer is missing from the PATH")
+
+
+def main():
+    """The main function to run tests from the command line."""
+    args = parse_args()
+
+    # Check environment
+    environment_check(args.valgrind, args.wine)
+
+    jobs = abs(args.jobs or 1)
+    print(f'Running in {jobs} job(s)')
+
+    if not _ctx.re2c or not os.access(_ctx.re2c, os.X_OK):
+        raise RuntimeError(f'Cannot find re2c executable')
+
+    test_blddir = create_tests_tree()
+    copy_tests(args.tests, test_blddir)
+
+    if os.name == 'posix':
+        subprocess.call(['chmod', '-R', 'u+w', test_blddir])
+
+    clean_test_tree(test_blddir)
+    tests = filter_tests(test_blddir)
+
+    with multiprocessing.Pool(
+            processes=jobs,
+            initializer=init_process,
+            # Note: the arguments order is important here
+            initargs=(test_blddir, args.skeleton, args.keep_temp_files,
+                      args.valgrind, args.wine,)
+    ) as pool:
+        pool.map(run_pack, chunks(tests, int(len(tests) / jobs + 1)))
+
+    # collect results from all threads
+    total = len(tests)
+    total_ran_tests = 0
+    total_hard_errors = 0
+    total_soft_errors = 0
+
+    for log in glob_re(r'\d+_\d+\.log$', test_blddir):
+        log = abshere(test_blddir, log)
+        total_ran_tests += grep_count(log, r'ran[^\d]+')
+        total_hard_errors += grep_count(log, r'hard[^\d]+')
+        total_soft_errors += grep_count(log, r'soft[^\d]+')
+        remove(log)
+
+    # remove directories that are empty or contain only .inc, .h and .go files
+    for d in filter(lambda x: os.path.isdir(str(x)), Path(test_blddir).rglob('*')):
+        exts = ('.inc', '.h', '.go')
+        found = filter(lambda x: not str(x).endswith(exts), Path(d).rglob('*'))
+
+        for f in found:
+            if os.path.isdir(f):
+                shutil.rmtree(f, ignore_errors=True)
+            else:
+                remove(f)
```

Comment:
In case you want to rewrite it, there is some caution needed to avoid deleting too many or too few things. I check it by making some changes to the source code (e.g. change a couple of symbols in src/codegen/render.cc) to make tests fail in a "natural" way, and then make sure the number of reported failures is the same as `find . -name '*.re' | wc -l` in the test directory.

Reply:
I sent you a PR https://github.com/sergeyklay/re2c/pull/2 with this commit --- let me know if there is some easier way to add commits on top of an existing PR.

Reply:
@skvadrik I just changed this (a minute before your PR :) )